File: jlpt/native_ios/Sources/Data/DTOs/ContentDTOs.swift
import Foundation

struct VocabularyResponseDTO: Decodable {
    let id: Int64
    let level: JlptLevel
    let word: String
    let reading: String
    let meaningVi: String
    let meaningEn: String?
    let partOfSpeech: String?
    let exampleSentence: String?
    let exampleReading: String?
    let exampleMeaning: String?
}

struct KanjiResponseDTO: Decodable {
    let id: Int64
    let level: JlptLevel
    let character: String
    let onyomi: String?
    let kunyomi: String?
    let meaningVi: String
    let strokeCount: Int?
    let exampleWords: String?
}

struct GrammarResponseDTO: Decodable {
    let id: Int64
    let level: JlptLevel
    let pattern: String
    let meaningVi: String
    let meaningEn: String?
    let usageNote: String?
    let exampleSentence: String?
    let exampleMeaning: String?
}

struct PassageResponseDTO: Decodable {
    let id: Int64
    let level: JlptLevel
    let title: String
    let content: String
}

struct ListeningAudioResponseDTO: Decodable {
    let id: Int64
    let level: JlptLevel
    let title: String
    let audioUrl: String
    let transcript: String?
    let durationSeconds: Int?
}

/// Mirrors the subset of Spring Data's `Page<T>` JSON shape we actually use; unrecognized fields
/// (`pageable`, `sort`, `first`, `last`, ...) are ignored by `JSONDecoder` automatically.
struct PageDTO<T: Decodable>: Decodable {
    let content: [T]
    let totalElements: Int64?
    let totalPages: Int?
    let number: Int?
    let size: Int?
}

File: jlpt/native_android/app/src/main/java/com/jlpt/android/data/remote/dto/ContentDtos.kt
package com.jlpt.android.data.remote.dto

import com.jlpt.android.domain.model.JlptLevel
import kotlinx.serialization.Serializable

@Serializable
data class VocabularyResponseDto(
    val id: Long,
    val level: JlptLevel,
    val word: String,
    val reading: String,
    val meaningVi: String,
    val meaningEn: String? = null,
    val partOfSpeech: String? = null,
    val exampleSentence: String? = null,
    val exampleReading: String? = null,
    val exampleMeaning: String? = null
)

@Serializable
data class KanjiResponseDto(
    val id: Long,
    val level: JlptLevel,
    val character: String,
    val onyomi: String? = null,
    val kunyomi: String? = null,
    val meaningVi: String,
    val strokeCount: Int? = null,
    val exampleWords: String? = null
)

@Serializable
data class GrammarResponseDto(
    val id: Long,
    val level: JlptLevel,
    val pattern: String,
    val meaningVi: String,
    val meaningEn: String? = null,
    val usageNote: String? = null,
    val exampleSentence: String? = null,
    val exampleMeaning: String? = null
)

@Serializable
data class PassageResponseDto(
    val id: Long,
    val level: JlptLevel,
    val title: String,
    val content: String
)

@Serializable
data class ListeningAudioResponseDto(
    val id: Long,
    val level: JlptLevel,
    val title: String,
    val audioUrl: String,
    val transcript: String? = null,
    val durationSeconds: Int? = null
)

/** Mirrors the subset of Spring Data's `Page<T>` JSON shape we actually use. */
@Serializable
data class PageDto<T>(
    val content: List<T>,
    val totalElements: Long = 0,
    val totalPages: Int = 0,
    val number: Int = 0,
    val size: Int = 0
)
